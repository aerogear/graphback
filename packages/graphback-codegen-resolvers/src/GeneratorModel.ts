
/**
 * Defines generator output
 */
export interface OutputFileSystem {
    resolvers: FileDefinition
    index: FileDefinition
}

/**
 * Contains definition of the single file
 */
export interface FileDefinition {
    fileName: string
    output: string
}