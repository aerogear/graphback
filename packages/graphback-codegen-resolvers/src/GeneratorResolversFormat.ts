
/**
 * Format used to agregate resolvers
 */
export interface GeneratorResolversFormat { Query: {}, Mutation: {}, Subscription: {}};
