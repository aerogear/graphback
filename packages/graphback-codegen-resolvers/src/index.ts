export * from "./api/OutputResolver"
export * from "./api/generateResolvers"
export * from './api/ResolverGeneratorOptions'