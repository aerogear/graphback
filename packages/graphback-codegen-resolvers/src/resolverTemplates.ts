export const createTemplate = (modelName: string): string => {
  return `(parent, args, context) => {
        return context.${modelName}.create(args, context);
    }`;
}

export const updateTemplate = (modelName: string): string => {
  return `(parent, args, context) => {
      return context.${modelName}.update(args, context);
    }`
}

export const deleteTemplate = (modelName: string): string => {
  return `(parent, args, context) => {
      return context.${modelName}.delete(args, context);
    }`
}

export const findAllTemplate = (modelName: string): string => {
  return `(parent, args, context) => {
      return context.${modelName}.findAll();
    }`
}

export const findTemplate = (modelName: string): string => {
  return `(parent, args, context) => {
      return context.${modelName}.findBy(args);
    }`
}

export const newSubscriptionTemplate = (modelName: string): string => {
  return `{
      subscribe: (parent, args, context) => {
        return context.${modelName}.subscribeToCreate(args, context);
      }
    }`
}

export const updatedSubscriptionTemplate = (modelName: string): string => {
  return `{
      subscribe: (parent, args, context) => {
        return context.${modelName}.subscribeToUpdate(args, context);
      }
    }`
}

export const deletedSubscriptionTemplate = (modelName: string): string => {
  return `{
      subscribe: (parent, args, context) => {
        return context.${modelName}.subscribeToDelete(args, context);
      }
    }`
}

export const blankResolver = `(parent, args, context) => {
    // Implementation here
}`;

export const blankSubscription = `{
    subscribe: (parent, args, context) => {
        // Implementation here
    }
}`
