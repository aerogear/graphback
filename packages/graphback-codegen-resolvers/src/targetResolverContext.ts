import { getFieldName, getTableName, GraphbackOperationType, InputModelFieldContext, InputModelTypeContext, OBJECT_TYPE_DEFINITION } from "@graphback/codegen-core";;
import { ResolverRelationContext, ResolverTypeContext, TargetResolverContext } from './api/resolverTypes';
import * as serviceTemplates from './templates/LayeredResolverTemplates';

 
/**
 * Resolver crud methods - create, update, delete, find and findAll
 */
const createResolver = (t: InputModelTypeContext, database: string): string => {
  if (t.config.create) {
    return serviceTemplates.createTemplate(database, t.config.subCreate, getFieldName(t.name, GraphbackOperationType.CREATE), getTableName(t.name), t.name)
  }

  return undefined
}

const updateResolver = (t: InputModelTypeContext): string => {
  if (t.config.update) {
    return serviceTemplates.updateTemplate(t.config.subUpdate, getFieldName(t.name, GraphbackOperationType.UPDATE), getTableName(t.name), t.name)
  }

  return undefined
}

const deleteResolver = (t: InputModelTypeContext): string => {
  if (t.config.delete) {
    return serviceTemplates.deleteTemplate(t.config.subDelete, getFieldName(t.name, GraphbackOperationType.DELETE), getTableName(t.name), t.name)
  }

  return undefined
}

const findResolver = (t: InputModelTypeContext): string => {
  if (t.config.find) {
    return serviceTemplates.findTemplate(getFieldName(t.name, GraphbackOperationType.FIND, 's'), getTableName(t.name))
  }

  return undefined
}

const findAllResolver = (t: InputModelTypeContext): string => {
  if (t.config.findAll) {
    return serviceTemplates.findAllTemplate(getFieldName(t.name, GraphbackOperationType.FIND_ALL, 's'), getTableName(t.name))
  }

  return undefined
}

/**
 * Subscriptions - new, updated and deleted
 */

const newSub = (t: InputModelTypeContext): string => {
  if (t.config.create && t.config.subCreate) {
    return serviceTemplates.newSub(t.name)
  }

  return undefined
}

const updatedSub = (t: InputModelTypeContext): string => {
  if (t.config.update && t.config.subUpdate) {
    return serviceTemplates.updatedSub(t.name)
  }

  return undefined
}

const deletedSub = (t: InputModelTypeContext): string => {
  if (t.config.delete && t.config.subDelete) {
    return serviceTemplates.deletedSub(t.name)
  }

  return undefined
}

/**
 * Create context object for each individual type
 * @param context Visited info from the model
 */
export const buildGraphbackOperationTypeContext = (context: InputModelTypeContext, database: string, relations: string[]): TargetResolverContext => {
  const typeContext = {
    relations: [],
    queries: [],
    mutations: [],
    subscriptions: [],
  }

  if (relations.length) {
    typeContext.relations = relations
  }
  if (!context.config.disableGen) {
    typeContext.queries = [findResolver(context), findAllResolver(context)].filter((s: string) => s !== undefined)
    typeContext.mutations = [createResolver(context, database), updateResolver(context), deleteResolver(context)].filter((s: string) => s !== undefined)
    typeContext.subscriptions = [newSub(context), updatedSub(context), deletedSub(context)].filter((s: string) => s !== undefined)
  }

  return typeContext
}

/**
 * Create context of all the types
 * @param input Input visited object
 */
export const buildResolverTargetContext = (input: InputModelTypeContext[], database: string) => {
  const inputContext = input.filter((t: InputModelTypeContext) => t.kind === OBJECT_TYPE_DEFINITION && t.name !== 'Query' && t.name !== 'Mutation' && t.name !== 'Subscription')
  const output: ResolverTypeContext[] = []

  const relations = []

  inputContext.forEach((t: InputModelTypeContext) => {
    t.fields.forEach((f: InputModelFieldContext) => {
      if (f.isType) {
        if (f.directives.OneToOne || !f.isArray) {
          let columnName = `${t.name.toLowerCase()}Id`
          if (f.directives.OneToOne) {
            columnName = f.directives.OneToOne.field
          }
          relations.push({
            typeName: t.name,
            implementation: serviceTemplates.typeRelation('OneToOne', columnName, f.name, f.type.toLowerCase())
          })
        } else if (f.directives.OneToMany || f.isArray) {
          let columnName = `${t.name.toLowerCase()}Id`
          if (f.directives.OneToMany) {
            columnName = f.directives.OneToMany.field
          }
          relations.push({
            typeName: t.name,
            implementation: serviceTemplates.typeRelation('OneToMany', columnName, f.name, f.type.toLowerCase())
          })
        }
      }
    })
  })

  inputContext.forEach((t: InputModelTypeContext) => {
    output.push({
      name: t.name,
      context: buildGraphbackOperationTypeContext(t, database, relations.filter((r: ResolverRelationContext) => r.typeName === t.name).map((r: ResolverRelationContext) => r.implementation))
    })
  })

  return output
}

/**
 * Create queries, mutations or subscriptions from custom input provided
 * @param inputContext Input visited object
 */
export const createCustomContext = (inputContext: InputModelTypeContext[]) => {
  const queryType = inputContext.filter((t: InputModelTypeContext) => t.name === 'Query')
  let customQueries = []
  if (queryType.length) {
    customQueries = queryType[0].fields.map((f: InputModelFieldContext) => {
      return {
        name: f.name,
        implementation: serviceTemplates.blankResolver(f.name),
        operationType: 'Query'
      }
    })
  }

  const mutationType = inputContext.filter((t: InputModelTypeContext) => t.name === 'Mutation')
  let customMutations = []
  if (mutationType.length) {
    customMutations = mutationType[0].fields.map((f: InputModelFieldContext) => {
      return {
        name: f.name,
        implementation: serviceTemplates.blankResolver(f.name),
        operationType: 'Mutation'
      }
    })
  }

  const subscriptionType = inputContext.filter((t: InputModelTypeContext) => t.name === 'Subscription')
  let customSubscriptions = []
  if (subscriptionType.length) {
    customSubscriptions = subscriptionType[0].fields.map((f: InputModelFieldContext) => {
      return {
        name: f.name,
        implementation: serviceTemplates.blankSubscription(f.name),
        operationType: 'Subscription'
      }
    })
  }

  return [...customQueries, ...customMutations, ...customSubscriptions]
}
