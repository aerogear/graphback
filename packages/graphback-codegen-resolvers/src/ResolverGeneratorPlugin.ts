import { GraphbackCoreMetadata, GraphbackPlugin } from '@graphback/core';
import { OutputFileSystem } from './GeneratorModel';
import { GeneratorResolversFormat } from './GeneratorResolversFormat';
import { generateCRUDResolversFunctions } from './templates/createResolvers';
import { createResolverTemplate } from './templates/resolverWrapper';
import { createModelsFile } from './templates/modelTemplate';
import { writeResolvers } from './writeResolvers';

export interface ResolverGeneratorPluginConfig {
    /**
     * Extension of the generated files 
     */
    format: 'ts' | 'js'

    /**
     * RelativePath for the output files created by generator
     */
    outputPath: string

    /**
     * Name of the generated resolvers file (default: resolvers.(format))
     */
    resolversFileName?: string

    /**
     * Layout of the of the resolvers object. 
     * Supports Apollo (GraphQL-Tools) or GraphQL reference spec format
     */
    layout?: 'apollo' | 'graphql'
}

const PLUGIN_NAME = 'CRUD_RESOLVER_GENERATOR';

/**
 * Graphback Resolver File Generator Plugin
 * 
 * Generates:
 * 
 * - default CRUD resolvers for all model types.
 * - blank resolver files to implements custom resolvers
 * - CRUD Services for each type for controling the data sources
 * 
 * And then writes to content to resolver files in the server.
 * 
 */
export class ResolverGeneratorPlugin extends GraphbackPlugin {

    private pluginConfig: ResolverGeneratorPluginConfig;

    public constructor(pluginConfig: ResolverGeneratorPluginConfig) {
        super();
        this.pluginConfig = Object.assign({
            format: 'ts',
            layout: "apollo",
            resolversFileName: 'resolvers',
        }, pluginConfig);
        if (!pluginConfig.outputPath) {
            throw new Error("resolver plugin requires outputPath parameter")
        }
    }

    public getPluginName() {
        return PLUGIN_NAME;
    }

    public createResources(metadata: GraphbackCoreMetadata): void {
        const outputResolves = this.generateResolvers(metadata);
        if (outputResolves) {
            writeResolvers(outputResolves, this.pluginConfig);
        }
    }

    public generateResolvers(metadata: GraphbackCoreMetadata): OutputFileSystem {
        const models = metadata.getModelDefinitions();
        if (models.length === 0) {
            this.logWarning("Provided schema has no models. Cannot generate resolvers")

            return undefined;
        }

        const generatedResolversFunctions = generateCRUDResolversFunctions(models);
        const generatedResolverFile = this.createGeneratedResolversFile(generatedResolversFunctions, this.pluginConfig);
        const contextFile = createModelsFile(models, this.pluginConfig);

        return {
            resolvers: generatedResolverFile,
            context: contextFile
        };
    }

    protected createGeneratedResolversFile(resolvers: GeneratorResolversFormat, pluginConfig: ResolverGeneratorPluginConfig) {
        const generatedResolvers = createResolverTemplate(resolvers, pluginConfig);

        return {
            fileName: `${pluginConfig.resolversFileName}.${pluginConfig.format}`,
            output: generatedResolvers
        };
    }
}