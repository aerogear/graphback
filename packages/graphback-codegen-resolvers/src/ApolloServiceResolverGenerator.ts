import { InputModelTypeContext } from "@graphback/core"
import { ResolverTypeContext } from './api/resolverTypes';
import { buildResolverTargetContext } from './api/targetResolverContext';
import { generateGraphbackResolvers } from './formatters/ApolloResolverFormatter';

/**
 * Generate resolvers layer that connects with service
 */
export class ApolloServiceResolverGenerator {
    private context: ResolverTypeContext[]
    private inputContext: InputModelTypeContext[]

    constructor(inputContext: InputModelTypeContext[]) {
      this.inputContext = inputContext
    }

    public generate(database: string) {
      this.context = buildResolverTargetContext(this.inputContext)

      return generateGraphbackResolvers(this.context, this.inputContext)
    }
  }
