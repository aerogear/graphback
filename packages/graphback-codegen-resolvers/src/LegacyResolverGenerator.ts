<<<<<<< HEAD:packages/graphback-codegen-resolvers/src/legacy-knex/LegacyResolverGenerator.ts
import { InputModelTypeContext } from "@graphback/core"
import { generateGraphbackResolvers } from './LegacyApolloResolverTemplate';
import { ResolverTypeContext } from './resolverTypes';
import { buildResolverTargetContext, createCustomContext } from './targetResolverContext';
=======
import { InputModelTypeContext } from "@graphback/codegen-core"
import { ResolverTypeContext } from './api/resolverTypes';
import { buildResolverTargetContext } from './targetResolverContext';
import { generateGraphbackResolvers } from './formatters/ApolloResolverFormatter';
>>>>>>> b2ebcd6... fix: remove legacy resolver generators:packages/graphback-codegen-resolvers/src/LegacyResolverGenerator.ts

/**
 * generate schema using context created using visitor pattern
 * and string templates
 */
export class LegacyResolverGenerator {
    private context: ResolverTypeContext[]
    private inputContext: InputModelTypeContext[]

    constructor(inputContext: InputModelTypeContext[]) {
      this.inputContext = inputContext
    }

    public generate(database: string) {
      this.context = buildResolverTargetContext(this.inputContext, database)

      return generateGraphbackResolvers(this.context, this.inputContext)
    }
  }
