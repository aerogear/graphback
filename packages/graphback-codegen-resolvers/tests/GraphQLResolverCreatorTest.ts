//tslint:disable-next-line: match-default-export-name no-implicit-dependencies
import { readFileSync } from 'fs';
import { GraphbackCoreMetadata } from '@graphback/core';
import { buildSchema } from 'graphql';
import { ResolverGeneratorPlugin } from '../src'

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

test('Test plugin engine ts', async () => {
  const crudMethods = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
  }

  const metadata = new GraphbackCoreMetadata({ crudMethods }, buildSchema(schemaText))
  const plugin = new ResolverGeneratorPlugin({ format: 'ts', outputPath: './tmp' });
  expect(plugin.generateResolvers(metadata)).toMatchSnapshot();
});


test('Test plugin engine js', async () => {
  const crudMethods = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
  }

  const metadata = new GraphbackCoreMetadata({ crudMethods }, buildSchema(schemaText))
  const plugin = new ResolverGeneratorPlugin({ format: 'js', outputPath: './tmp' });
  expect(plugin.generateResolvers(metadata)).toMatchSnapshot();
});