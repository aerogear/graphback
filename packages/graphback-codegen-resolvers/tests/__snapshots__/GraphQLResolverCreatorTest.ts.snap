// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test plugin engine js 1`] = `
Object {
  "context": Object {
    "fileName": "models.js",
    "output": "const models = [
  {
    name: \\"Note\\",
    pubSub: {
      publishCreate: true,
      publishUpdate: true,
      publishDelete: true,
    },
  },
  {
    name: \\"Comment\\",
    pubSub: {
      publishCreate: true,
      publishUpdate: true,
      publishDelete: true,
    },
  },
]

module.exports = { models }
",
  },
  "resolvers": Object {
    "fileName": "graphbackResolvers.js",
    "output": "/*
 * File generated by Graphback CRUD resolver plugin.
 * Content will be overwritten by Graphback generator.
 * To change implementation please disable generator options and supply your own implementation
 * outside generated file.
 */

module.exports = {
  Note: {
    comments: (parent, { filter }, context) => {
      return context.Comment.batchLoadData(\\"noteId\\", parent.id, filter, context)
    },
  },

  Comment: {
    noteComment: (parent, args, context) => {
      return context.Note.findBy({ id: parent.noteId }).then(
        (results) => results[0],
      )
    },
  },

  Query: {
    findNote: (parent, args, context) => {
      return context.Note.findOne(args)
    },
    findNotes: (parent, { filter, page }, context) => {
      return context.Note.findBy(filter, page)
    },
    findAllNotes: (parent, { page }, context) => {
      return context.Note.findAll(page)
    },
    findComment: (parent, args, context) => {
      return context.Comment.findOne(args)
    },
    findComments: (parent, { filter, page }, context) => {
      return context.Comment.findBy(filter, page)
    },
    findAllComments: (parent, { page }, context) => {
      return context.Comment.findAll(page)
    },
  },

  Mutation: {
    createNote: (parent, args, context) => {
      return context.Note.create(args.input, context)
    },
    updateNote: (parent, args, context) => {
      return context.Note.update(args.input, context)
    },
    deleteNote: (parent, args, context) => {
      return context.Note.delete(args.input, context)
    },
    createComment: (parent, args, context) => {
      return context.Comment.create(args.input, context)
    },
    updateComment: (parent, args, context) => {
      return context.Comment.update(args.input, context)
    },
    deleteComment: (parent, args, context) => {
      return context.Comment.delete(args.input, context)
    },
  },

  Subscription: {
    newNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToCreate(args, context)
      },
    },
    updatedNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToUpdate(args, context)
      },
    },
    deletedNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToDelete(args, context)
      },
    },
    newComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToCreate(args, context)
      },
    },
    updatedComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToUpdate(args, context)
      },
    },
    deletedComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToDelete(args, context)
      },
    },
  },
}
",
  },
}
`;

exports[`Test plugin engine ts 1`] = `
Object {
  "context": Object {
    "fileName": "models.ts",
    "output": "export const models = [
  {
    name: \\"Note\\",
    pubSub: {
      publishCreate: true,
      publishUpdate: true,
      publishDelete: true,
    },
  },
  {
    name: \\"Comment\\",
    pubSub: {
      publishCreate: true,
      publishUpdate: true,
      publishDelete: true,
    },
  },
]
",
  },
  "resolvers": Object {
    "fileName": "graphbackResolvers.ts",
    "output": "/*
 * File generated by Graphback CRUD resolver plugin.
 * Content will be overwritten by Graphback generator.
 * To change implementation please disable generator options and supply your own implementation
 * outside generated file.
 */

export default {
  Note: {
    comments: (parent, { filter }, context) => {
      return context.Comment.batchLoadData(\\"noteId\\", parent.id, filter, context)
    },
  },

  Comment: {
    noteComment: (parent, args, context) => {
      return context.Note.findBy({ id: parent.noteId }).then(
        (results) => results[0],
      )
    },
  },

  Query: {
    findNote: (parent, args, context) => {
      return context.Note.findOne(args)
    },
    findNotes: (parent, { filter, page }, context) => {
      return context.Note.findBy(filter, page)
    },
    findAllNotes: (parent, { page }, context) => {
      return context.Note.findAll(page)
    },
    findComment: (parent, args, context) => {
      return context.Comment.findOne(args)
    },
    findComments: (parent, { filter, page }, context) => {
      return context.Comment.findBy(filter, page)
    },
    findAllComments: (parent, { page }, context) => {
      return context.Comment.findAll(page)
    },
  },

  Mutation: {
    createNote: (parent, args, context) => {
      return context.Note.create(args.input, context)
    },
    updateNote: (parent, args, context) => {
      return context.Note.update(args.input, context)
    },
    deleteNote: (parent, args, context) => {
      return context.Note.delete(args.input, context)
    },
    createComment: (parent, args, context) => {
      return context.Comment.create(args.input, context)
    },
    updateComment: (parent, args, context) => {
      return context.Comment.update(args.input, context)
    },
    deleteComment: (parent, args, context) => {
      return context.Comment.delete(args.input, context)
    },
  },

  Subscription: {
    newNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToCreate(args, context)
      },
    },
    updatedNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToUpdate(args, context)
      },
    },
    deletedNote: {
      subscribe: (parent, args, context) => {
        return context.Note.subscribeToDelete(args, context)
      },
    },
    newComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToCreate(args, context)
      },
    },
    updatedComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToUpdate(args, context)
      },
    },
    deletedComment: {
      subscribe: (parent, args, context) => {
        return context.Comment.subscribeToDelete(args, context)
      },
    },
  },
}
",
  },
}
`;
